### PR TITLE
support configuring fail point in RAII style

### PR DESCRIPTION
**Pull request by tabokie**
It is very easy to write failpoint tests without freeing up the failpoint in the end.

Signed-off-by: tabokie <xy.tao@outlook.com>

### CODE REVIEW

**Comment by tabokie** [comment]
@BusyJay What do you think

**Comment by BusyJay** [comment]
Why not use `FailScenario`?

**Comment by tabokie** [comment]
User might not want to configure a global hook that applies to all failpoint tests (like we do in TiKV). If they don't do that, they have to add additional two lines for every tests. `FailGuard` appears easier to use.

Also, with respect to TiKV's implementation, calling `teardown` after the test closure is executed is not safe. Some threads might already be joined (and blocked on failpoint) inside the test closure (via custom Drop implementation).

**Comment by BusyJay** [comment]
> User might not want to configure a global hook that applies to all failpoint tests (like we do in TiKV

Actually, TiKV configures global hook using its own way.

> I already used similar thing in Raft Engine (https://github.com/tikv/raft-engine/blob/1a9dd43a74a6ff93c176670e584db6ec520b6953/tests/failpoints/test_io_error.rs#L155), it's really natural to use.

Your usage is wrong. Technically, there is only one global failpoint registry, so without synchronization, failpoint configuration set by any cases will affect other cases as well. Hence executing cases in sequential order is required.

> Some threads might already be joined (and blocked on failpoint) inside the test closure (via custom Drop implementation).

It doesn't matter. Teardown will wake up all blocking threads.

**Comment by tabokie** [comment]
> Actually, TiKV configures global hook using its own way.

I'm saying user might **now** want to go over the trouble to doing a configuration like this.

> Your usage is wrong.

You are looking at the wrong thing. I'm demonstrating that using `FailGuard` can make **one** test case much easier to manage. The failpoint tests in Raft Engine is all run with one threads (and documented).

> It doesn't matter. Teardown will wake up all blocking threads.

It does matter. As I explained, the object destruction can be called inside the test closure. That means the process will be blocked inside without getting a chance to call the `teardown` outside the closure.

File: src/lib.rs [hunk 1]
```diff
@@ -708,6 +708,40 @@ pub fn remove<S: AsRef<str>>(name: S) {
...
+    pub fn new<S: Into<String>>(name: S, actions: &str) -> Result<FailGuard, String> {
+        let name = name.into();
+        cfg(&name, actions)?;
+        Ok(FailGuard(name))
+    }
+
+    /// Configure the actions for a fail point during the lifetime of the returning object.
+    ///
+    /// Read documentation of [`cfg_callback`] for more details.
+    pub fn new_callback<S, F>(name: S, f: F) -> Result<FailGuard, String>
```
Comment (BusyJay) [inline src/lib.rs:734]: Should be named `with_callback`.

**Review by BusyJay — COMMENTED** [review]
Rest LGTM

**Review by BusyJay — APPROVED** [review]

**Comment by tabokie** [comment]
@BusyJay I can't merge this.

**Pull request merged**